grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,314 @@
+            HashMultimap.create();
+    @Nullable
+    private Server server;
+
+    GrpcHealthCheckService(
+            Set<ListenableHealthChecker> serverHealthCheckers,
+            Map<String, ListenableHealthChecker> grpcServiceHealthCheckers,
+            List<HealthCheckUpdateListener> updateListeners
+    ) {
+        serverHealth = new SettableHealthChecker(false);
+        if (!updateListeners.isEmpty()) {
+            addServerHealthUpdateListener(updateListeners);
+        }
+        this.serverHealthCheckers = ImmutableSet.<ListenableHealthChecker>builder()
+                                                .add(serverHealth)
+                                                .addAll(serverHealthCheckers)
+                                                .build();
+        this.grpcServiceHealthCheckers = grpcServiceHealthCheckers;
+        final Consumer<String> healthCheckUpdateListener = provideInternalHealthUpdateListener();
+        setHealthUpdateListenerToServerHealthCheckers(healthCheckUpdateListener, this.serverHealthCheckers);
+        setHealthUpdateListenerToGrpcServiceHealthCheckers(healthCheckUpdateListener,
+                                                           this.grpcServiceHealthCheckers.entrySet());
+    }
+
+    @Override
+    public void check(HealthCheckRequest request, StreamObserver<HealthCheckResponse> responseObserver) {
+        final String service = request.getService();
+        final ServingStatus status = checkServingStatus(service);
+        if (status == ServingStatus.SERVICE_UNKNOWN) {
+            responseObserver.onError(Status.NOT_FOUND
+                                             .withDescription(String.format(
+                                                     "The service name(%s) is not registered in this service",
+                                                     service))
+                                             .asRuntimeException());
+            return;
+        }
+        responseObserver.onNext(getHealthCheckResponse(status));
+        responseObserver.onCompleted();
+    }
+
+    @Override
+    public void watch(HealthCheckRequest request, StreamObserver<HealthCheckResponse> responseObserver) {
+        ServiceRequestContext.current().clearRequestTimeout();
+        final String service = request.getService();
+        synchronized (watchers) {
+            final ServingStatus status = checkServingStatus(service);
+            final HealthCheckResponse response = getHealthCheckResponse(status);
+            responseObserver.onNext(response);
+            watchers.put(service, Maps.immutableEntry(responseObserver, response.getStatus()));
+        }
+        ((ServerCallStreamObserver<HealthCheckResponse>) responseObserver).setOnCancelHandler(() -> {
+            synchronized (watchers) {
+                final ServingStatus currentStatus = checkServingStatus(service);
+                watchers.get(service).remove(Maps.immutableEntry(responseObserver, currentStatus));
```

Comment:
What if `currentStatus` is different from the last updated status?
Don't we need to iterate all elements to find the `responseObserver`?

Reply:
Agree with this comment.

I also think modifying `watchers` to be a map of maps is not a bad idea either

i.e.
`Map<String, Map<StreamObserver<HealthCheckResponse>, ServingStatus>>` where the inner map is an `IdentityHashMap`

Reply:
One more thing we have to think about is that: should we keep the `currentStatus`?
It's perfectly valid to notify the same status and I think notifying the same status shouldn't occur that much.
The upstream also does not keep the `currentStatus`: https://github.com/grpc/grpc-java/blob/bab1fe38dc50d4178955b79cfb1636afd2aa64df/services/src/main/java/io/grpc/protobuf/services/HealthServiceImpl.java#L179

Reply:
> A client can call the Watch method to perform a streaming health-check. The server will immediately send back a message indicating the current serving status. It will then subsequently send a new message whenever the service's serving status changes.

I don't see any statement that specifies a message should be sent "if and only if" the status is changed.

> One more thing we have to think about is that: should we keep the currentStatus?

If it helps keep our implementation simpler, I agree 😄

Reply:
I think your opinion makes more sense.
I'm going to rewrite the current implementation that holds `currentStatus`.